### PR TITLE
fix(service-bot): drive restart completion with a durable poll

### PR DESCRIPTION
**Pull request by totalfrank**
## Problem

A restart is durably **submitted** but never durably **observed**.

`restart_bot` → `enqueue_restart` → `PublishRestartHandler` → `execute_restart` is crash-safe end to end (the operation runner adopts an in-doubt workflow instead of issuing a second one, and completes the op inline). But every piece of work the restart still owes afterwards lives in `sync_restart_progress`, whose only caller is the user-facing `/restart_status` endpoint. A backend invariant is therefore gated on a client choosing to poll.

With no client polling, four things never happen:

1. `ext.restart.restarting` is never cleared — the record reads "restarting" forever.
2. On the recreate leg (target bot gone, or a non-`UPGRADE` online decision) the freshly minted binding stays **PENDING**; only the sync activates it. As the code comment there says, it "stays PENDING forever and binding consumers reject it" — the bot is deployed but unusable. This is the one with real user impact.
3. A recreated teclaw container never gets its post-deploy MCP outbound/auth rule (`_refresh_provider_mcp_after_success`).
4. A BaaS-side restart failure is never recorded: the record is not marked FAILED, and the ledger's `fail_by_workflow` outcome correction never runs — so a failed deploy keeps reading as the live deployment, and `is_current_online_deployment` will let the *next* online release skip re-issuing on the strength of it.

Two entry points reach this. The `/restart` endpoint is the obvious one. The second is `retry()` when the rollback target is `SUCCESS`: `_retry_via_restart` runs `execute_restart` inline and enqueues `progress_poll`, which completes immediately because `SUCCESS` is outside `_POLL_ACTIVE_STATES` — the comment there named `/restart_status` as the thing that settles it.

## Solution

Add a durable poll, `service_bot.publish.restart_poll`.

Reusing `progress_poll` does not work: it keys **both** its wait state and its terminal condition on the publish status, and a restart runs on a stable record (`VALIDATING` / `SUCCESS`) that `execute_restart` deliberately never transitions. It would short-circuit on `_POLL_ACTIVE_STATES` and complete without calling the sync at all.

So `PublishRestartPollHandler` swaps both judgements:

- **wait state** → `ext.restart.restarting`, via `is_restart_in_progress`, which also requires the record to still be in a status `sync_restart_progress` can resolve a stage for. (That function derives its stage from the status too, so a record outside the restartable statuses cannot be reconciled by it either — the guard turns a 24h reschedule spin into a clean exit rather than costing a reconciliation.)
- **terminal signal** → the BaaS restart workflow's own status. `SUCCESS` completes; `FAILED` fails the task (the sync has already marked the record, so the task row stays semantically honest); anything else — including the sync's no-data early returns, where a transient progress-fetch error must retry — reschedules on the same 8s cadence and 24h give-up horizon as the publish poll.

It is enqueued from both entry points:

- `PublishRestartHandler`, after `execute_restart` reports success. That placement is deliberate rather than beside `enqueue_restart`: by then the workflow id is recorded in both the ledger and `ext.restart.<stage>`, so the poll cannot read the marker before the restart wrote it — the same-batch race that forces `_retry_via_restart` to run its restart inline — and one call site covers all three of `execute_restart`'s success exits.
- `_retry_via_restart`, alongside the existing `enqueue_progress_poll`. Keying on the marker rather than the status means it covers both rollback targets. For the `*_PUB` ones that leaves two pollers on one idempotent sync; whichever observes the terminal workflow first ends both.

### Idempotency at the handoff

`execute_restart` COMPLETEs its ledger op before returning, and `open_publish_operation` resumes only a *non-terminal* op — past a terminal one it opens the next attempt. So an at-least-once redelivery would issue a **second BaaS restart**. `has_unreconciled_restart` recognises a restart the record already submitted (marker set, workflow handle recorded, and no deploy operation still open for the stage), and the handler then re-runs only the poll handoff.

The third condition is load-bearing: a crash mid-`execute_restart`, or between a recreate's ext write and its completion, leaves a RESTART or FIRST_RELEASE op open, and those must fall *through* the guard so the runner resumes and finalises them as it does today.

That guard is what lets `enqueue_restart_poll` propagate its failures rather than swallow them, per `AGENTS.md:264-265` — a failed write retries the task, which re-enqueues the poll without re-deploying.

> ⚠️ **Behaviour change worth a reviewer's eye:** a second `/restart` arriving while one is genuinely in flight is now absorbed as a redelivery instead of issuing a second restart. The endpoint's response is unchanged. I believe this is the better behaviour, but it is a real change and should not slip through unremarked.

The restart handler's failure branch deliberately leaves the marker **untouched**. Every `"success": False` return in `execute_restart` is a preflight check that runs before the marker is written at `restart_mixin.py:238`; past that write the function only returns success or raises. So that branch is reachable only when this execution never set the marker, and anything present belongs to a concurrent restart.

Both new accessors live on `RestartMixin` with the rest of the restart concern, and are deliberately **not** on `PublishFlowServiceProtocol` — that carries the API-boundary methods adapters call, not task-handler accessors (`get_publish_record` and `is_current_online_deployment` are absent for the same reason). So this is not a Service API contract change.

*(Everything under this heading, plus the marker paragraph, came out of Codex review rounds 1–5 — see the resolved threads.)*

Rejected alternative: folding both polls into one handler parameterised by (wait-state predicate, sync function, terminal reader). The shared skeleton is ~6 lines and all three semantically loaded parts differ; the file's convention is also one explicit handler class per task type, since `task_type` is the registry key.

## Validation

- **Full backend suite: 10359 passed, 3 skipped.**
- `test_publish_tasks.py` — 11 new tests covering the poll (enqueue-on-success, reschedule while pending, complete on `SUCCESS`, fail on `FAILED`, early exit when the marker is cleared, reschedule on no data, missing record) and the handoff (enqueue failure propagates, redelivery re-runs only the handoff, first delivery still issues, marker preserved on a failed preflight, the concurrent restart's poll still reaches the sync).
- `test_publish_flow_service.py` — 1 new test (a `SUCCESS` rollback target enqueues the restart poll) and 1 extended (a `VALIDATE_PUB` target enqueues both).
- `tests/community/architecture` — 112 passed, including the module-size gate below.
- SAST block-rule scan (the pre-push lint gate's rule set) over every changed file — clean, exit 0.

CI caught one thing local runs missed: `test_no_oversized_modules` failed because `publish_flow_service.py` had reached 1006 lines against a 1000 cap. Moving both accessors to `restart_mixin.py` returns it to **988** — exactly its length on the base branch.

## Compatibility and risk

Additive — a new task type plus two accessors on `RestartMixin`; no schema, protocol, contract or config change. `/restart_status` keeps working unchanged and remains the UI's live-progress source; the poll makes it redundant for correctness only. Registration happens in `PublishTaskLifecycle.bootstrap()` alongside the existing handlers. The one behavioural change is the concurrent-`/restart` absorption called out above.

### Known limits — all tracked in #784

These are pre-existing in mechanism; this branch improves several of them without closing any completely, and none is a regression against the current behaviour.

- **Redelivery can still duplicate a restart.** `has_unreconciled_restart` keys on a *reconciliation* marker, whose lifetime is shorter than a delivery's. If the pod dies after the poll is enqueued but before the task row is marked SUCCEEDED, the poll can observe a fast terminal workflow and clear the marker first — the redelivered task then finds no marker and issues a second BaaS restart. The guard narrows the pre-existing window (redelivery re-issues only once the marker is gone, instead of always) but cannot close it: the only sound discriminator is a delivery-carried operation identity, which is a change to the restart entry point.
- **Handoff enqueues are one-shot writes.** The `SUCCESS`-retry path runs `execute_restart` inline, so a failed poll enqueue has no durable row to redeliver, and the record has already left `FAILED` so `retry()` is closed to it. The same shape holds for `_retry_via_online_release`, `_retry_via_verify_flow` and `process`; the general fix is a transactional outbox.
- **Overlapping restarts share one marker**, so one restart's terminal sync can clear another's wait state.
- **A record superseded to `UPGRADED` mid-restart** can no longer be reconciled by `sync_restart_progress` at all, since that function derives its stage from the publish status.
- **Duplicate poll rows** are possible on a lease-expiry re-claim (the queue has no insert-time dedup). Tolerated rather than prevented: everything behind the sync is idempotent, and `verify_flow`/`online_release` already re-enqueue on every re-run.

Rollback is reverting the commits — the new task type has no persisted state of its own.

### CODE REVIEW

File: src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py [hunk 1]
```diff
@@ -375,7 +395,21 @@ async def _run(self, publish_id: int, stage: str, operator: str) -> TaskOutcome:
...
         if not result or not result.get("success"):
             message = (result or {}).get("message", "unknown error")
+            # execute_restart sets ext.restart.restarting *before* it issues, and
+            # only the sync clears it — so a restart that never got issued would
+            # leave the record reading "restarting" forever. Clear it here,
+            # mirroring retry's own cleanup of its restart flag on a failed
+            # submit. (A raised exception deliberately does NOT clear it: the task
+            # retries and the same op resumes, so the restart really is still in
+            # flight.)
+            self._flow.clear_restart_in_progress(publish_id)
```
[the remark this answers is not in the document]
Comment (totalfrank) [reply src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py]: Confirmed and fixed in `0670593`.

Verifying the mechanism against the source: `execute_restart` writes the marker at `restart_mixin.py:238`, and every `"success": False` return in that function sits at lines 210, 215, 218, 225 and 230 — all *before* it. Past line 238 the function only returns `"success": True` (261, 416, 553) or raises. So the failure branch of `PublishRestartHandler` is reachable **only** when this execution never set the marker, which makes the clear pure collateral damage: as you note, `restart_bot` doesn't reject a restart while one is in flight, so what it erased could be a concurrent restart's marker — and `is_restart_in_progress()` returning false then turns that restart's poll into an immediate no-op, stranding exactly the recreate-leg PENDING binding and unreconciled failures this task was added to prevent.

The fix is to drop the clear rather than to associate the marker with an operation. Given the invariant above there is never a marker of *ours* to clean up on that path, so scoping it per-operation would add state to guard a case that cannot arise. The unused `clear_restart_in_progress` accessor and its protocol stub are removed with it.

Two regression tests cover the scenario: one asserts the marker survives a failed preflight, the other that the concurrent restart's poll still reaches `sync_restart_progress` afterwards.

One thing I did **not** change: `restart_bot` still accepts a restart while one is in flight. That's pre-existing behaviour, concurrent restarts were possible before this PR, and adding a rejection would change the endpoint's contract — it belongs in its own change rather than here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py [hunk 2]
```diff
@@ -375,7 +395,23 @@ async def _run(self, publish_id: int, stage: str, operator: str) -> TaskOutcome:
...
+            # restart's marker (and its poll's wait state) can be sitting there.
+            # Clearing it would strand that restart exactly the way this task
+            # exists to prevent.
             return Fail(f"restart failed: publish_id={publish_id}, {message}")
+
+        # Enqueue the poll HERE rather than alongside ``enqueue_restart``: by now
+        # the workflow id is recorded (ledger + ext.restart.<stage>), so the poll
+        # cannot read ext.restart before the restart wrote it — the same-batch
+        # race that forces ``_retry_via_restart`` to run its restart inline.
+        enqueue_restart_poll(self._task_queue_service, publish_id=publish_id)
```
[the remark this answers is not in the document]
Comment (totalfrank) [reply src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py]: Confirmed and fixed in `527b9c9`.

The existing comment at `retry_ops_mixin.py:211-212` said it outright — *"SUCCESS: not a poll wait state — the poll completes immediately and the restart is settled by the user-called /restart_status endpoint."* That is exactly the dependency this branch exists to remove, reached through a second entry point I had missed.

`_retry_via_restart` now enqueues `enqueue_restart_poll` alongside the existing `enqueue_progress_poll`. Since the restart poll keys on `ext.restart.restarting` (set by the inline `execute_restart` above it) rather than on the publish status, it covers both rollback targets rather than just the `SUCCESS` one.

I added it rather than substituting it for `progress_poll`, so the proven `*_PUB` path — the `ext.retry` redirect into `sync_restart_progress` — is untouched. The cost is two pollers on one sync for `*_PUB` retries. That is harmless: `sync_restart_progress` is idempotent, and whichever observes the terminal workflow first clears the marker and advances the status, after which the other exits on its own guard. Noted in the comment so it reads as deliberate.

Covered by a new test asserting the `SUCCESS` rollback target enqueues the restart poll, plus the existing `VALIDATE_PUB` test extended to assert both tasks are enqueued.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/backend/src/agentclaw/community/api/publish_flow_service.py [hunk 3]
```diff
@@ -22,6 +22,8 @@ def sync_scale_progress(self, *args: Any, **kwargs: Any) -> Any: ...
 
     def sync_restart_progress(self, *args: Any, **kwargs: Any) -> Any: ...
 
+    def is_restart_in_progress(self, *args: Any, **kwargs: Any) -> Any: ...
```
[the remark this answers is not in the document]
Comment (totalfrank) [reply src/backend/src/agentclaw/community/api/publish_flow_service.py]: Agreed, and taken the first of your two options in `527b9c9` — `is_restart_in_progress` is out of the protocol, which reverts `api/publish_flow_service.py` to unmodified and drops it from the diff.

Keeping it out is the right call rather than the lesser one, because the protocol's own contents already draw this line. `PublishFlowServiceProtocol` carries the API-boundary methods that adapters call, and deliberately omits the task-handler accessors — `get_publish_record` and `is_current_online_deployment` both live in the same "public accessors for the durable task handlers" block on the concrete service and neither appears in the protocol. `is_restart_in_progress` is the third member of that block, and `tasks.py` types `_flow` as the concrete `PublishFlowService`, so the stub bought nothing at any call site.

So this stops being a contract change, and the AGENTS.md requirement you cited ("Contract changes require matching docs and conformance or compatibility tests") is satisfied by there no longer being one. Adding docs and a boundary test for a helper that no boundary consumer can reach would have widened the Service API surface to justify a line I should not have written.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (totalfrank) [reply src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py]: Confirmed, and the variant this PR introduced is fixed in `717b4e2`.

Verified the mechanism: `open_publish_operation` (`operation_runner.py:111-122`) returns the existing operation only while it is **non-terminal**; past a terminal one it takes `attempt + 1`, inserts a new operation, and `acquire_deploy_workflow` then issues a fresh BaaS restart. `execute_restart` calls `complete_operation` at `restart_mixin.py:409`, so by the time the poll enqueue runs the op is terminal and a redelivery would indeed restart the bot a second time.

The enqueue is now wrapped: on failure it logs and the handler still returns `Complete()`. That restart's completion falls back to `/restart_status` — the behaviour before this task existed — which is strictly better than re-deploying a bot. A test asserts the handler returns `Complete` rather than raising when the queue write fails.

What that does **not** fix is the crash window: a worker dying between `complete_operation` and the queue marking the row SUCCEEDED still redelivers. That window predates this PR — the handler already called `execute_restart` and returned `Complete()` with the same gap — so this change narrows it rather than opening it, and I have not tried to close it here. Doing so properly means giving the task a stable operation identity (`restart_bot` opening the ledger op up front, `reject_if_in_flight=True`, and the op id in the payload so `execute_restart` resumes that attempt), which is a change to the restart entry point and its contract.

Tracked with the full analysis in #784.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py [hunk 4]
```diff
@@ -513,6 +549,76 @@ def handle(self, payload: Optional[dict]) -> TaskOutcome:
...
+        publish_id = _require_int(payload, "publish_id")
+        record, _status = self._status(publish_id)
+        if record is None:
+            return Fail(f"publish record not found: publish_id={publish_id}")
+
+        # Already reconciled: a /restart_status call (or an earlier run of this
+        # task) observed the workflow terminal and cleared the marker, or the
+        # record left the statuses the sync can resolve a stage for.
+        if not self._flow.is_restart_in_progress(publish_id):
+            return Complete()
```
[the remark this answers is not in the document]
Comment (totalfrank) [reply src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py:619]: Real, verified, and deliberately **deferred** to #784 rather than fixed here. Flagging for the human reviewers so the decision is visible rather than buried.

The sequence holds as you describe it, and the consequence on B is exactly the failure this PR exists to prevent — a recreate leg's binding left PENDING, no MCP refresh, no FAILED marking.

Two things put the fix outside this PR:

**The mechanism predates it.** `_handle_sync_success` / `_handle_sync_failure` have always popped `restarting` without checking which workflow they reconciled, and `sync_restart_progress` has always resolved the workflow from the *latest* RESTART op rather than a caller-supplied id. This PR does not introduce the shared-boolean semantics; it gives that boolean a new consumer. Before it, the same overlap left the marker set and the newer restart equally unreconciled — the client-polling path had no better answer.

**Scoping it is not local.** Because the sync resolves the workflow itself, having the poll carry its own id is not sufficient — the sync's resolution rule is part of the same design. A correct fix stores the workflow id in the marker, threads it through the poll payload, and gates both clear sites on a match, which changes what `/restart_status` observes. That is a contract change on a shared endpoint, and `REL20260804` is a release branch.

Worth noting the cheaper root fix, which is why #784 covers this together with the redelivery hazard: adopting `reject_if_in_flight=True` at `restart_bot` — the runner's docstring calls it out as being "for user-triggered entry points that must reject a duplicate request" — makes overlapping restarts stop existing, and this hazard with them. That is also a contract change, which is exactly why it wants its own review rather than being folded in here.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by totalfrank** [comment]
## Review round 3 — status

Three rounds of automated review, five findings, all verified against the source before acting. Summary for the human reviewers, since one finding was deferred by decision rather than fixed.

| # | Round | Finding | Outcome |
|---|-------|---------|---------|
| 1 | 1 | P2 — failure branch cleared a concurrent restart's marker | **Fixed** `0670593` |
| 2 | 2 | P1 — restart poll missed retry-issued restarts | **Fixed** `527b9c9` |
| 3 | 2 | P1 — new Service API method lacked docs/conformance test | **Fixed** `527b9c9` (removed from the protocol; no contract change) |
| 4 | 3 | P1 — a poll-enqueue failure could redeliver the task and issue a second BaaS restart | **Fixed** `717b4e2` (introduced variant); residual crash window → #784 |
| 5 | 3 | P2 — shared `ext.restart.restarting` boolean lets one restart's sync clear another's wait state | **Deferred → #784** |

### On the deferral

Finding 5 is real and I reproduced the reasoning, but its mechanism predates this PR — `_handle_sync_success`/`_handle_sync_failure` have always cleared the marker without checking which workflow they reconciled, and `sync_restart_progress` has always resolved the workflow from the latest RESTART op rather than a caller-supplied id. This PR gives that boolean a new consumer; it does not create the semantics. A correct fix changes what `/restart_status` observes, which is a contract change on a shared endpoint, and this targets a release branch.

**#784** tracks it together with finding 4's residual crash window, since both have the same root: overlapping restarts are permitted, and the restart task has no stable operation identity. The issue notes the cheaper root fix — `reject_if_in_flight=True` at `restart_bot`, which the operation runner already supports for exactly this purpose — and why it wants its own review.

Both are reachable only when two restarts for one publish record overlap. Neither blocks the single-restart correctness this PR is about.

### ⚠️ CI has not run

No GitHub Actions run exists for this PR at any point — not on open, force-push, base change, ready-for-review, or any of the three fix pushes. `unit-tests`, `e2e-tests` and `singlebox-coverage` all trigger on `pull_request` against `'REL*'`, so the base branch is not the cause, and unrelated PRs completed runs throughout. `architecture-checks` correctly does not apply (it triggers on `main` only).

Standing in for it: **808 tests passing** across `core/service_bot`, `core/task_queue` and both durable-pipeline endpoint suites, plus a clean SAST block-rule scan over every changed file. **E2E, changed-line coverage and singlebox coverage are genuinely unrun** — please don't read the absence of red checks as green.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/backend/src/agentclaw/community/core/service_bot/services/publish_flow_service.py [hunk 5]
```diff
@@ -894,6 +894,24 @@ def get_publish_record(self, publish_id: int) -> BotPublishRecord | None:
...
+        The restart poll's wait-state test, standing in for the publish status
+        the restart path never transitions. The status half matters as much as
+        the marker: once a record leaves the restartable statuses the sync can
+        never reconcile it, so the poll must stop rather than spin to its
+        deadline."""
+        record = self._publish_service.get_publish_by_id(publish_id)
+        if record is None:
+            return False
+        if self._determine_restart_stage(PublishStatus(record.status)) is None:
+            return False
```
[the remark this answers is not in the document]
Comment (totalfrank) [reply src/backend/src/agentclaw/community/core/service_bot/services/publish_flow_service.py]: The scenario is real, but the status guard is not what causes it — so removing or reworking the guard alone would not reconcile anything.

`sync_restart_progress` derives the stage from the publish status itself (`progress_sync_mixin.py:554`) and early-returns when `_determine_restart_stage` yields `None`, with `"Current status … does not support querying restart progress"` and no `data`. `UPGRADED` is outside that mapping. So for a record `_mark_previous_publish_superseded` has moved to `UPGRADED`, the sync cannot resolve a stage regardless of what the poll does.

That means the two behaviours available today are:

- **with the guard** (current): the poll completes immediately;
- **without it**: `sync_restart_progress` returns no `data`, the poll takes its no-data branch and reschedules every 8s until the 24h deadline retires it as TIMED_OUT.

Neither reconciles the workflow. The guard converts a 10,800-iteration spin into a clean exit; it does not cost a reconciliation that would otherwise have happened.

Your suggested fix — deriving the stage from restart metadata or the task payload — is the right shape, but it has to be applied to **`sync_restart_progress`**, not just to the guard, because that function owns the stage resolution. That changes what `/restart_status` observes for a superseded record, which is a contract change on a shared endpoint and the same class of work as the marker-scoping in #784.

So: added to **#784** rather than fixed here, with the analysis above. Filed as its own item, since unlike the other two it does not require overlapping restarts — a single restart racing the next version's release is enough.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py [hunk 6]
```diff
@@ -375,7 +395,39 @@ async def _run(self, publish_id: int, stage: str, operator: str) -> TaskOutcome:
...
+        # its ledger op, and ``open_publish_operation`` resumes only a
+        # *non-terminal* op — past a terminal one it opens the next attempt. So a
+        # redelivery of this task would issue a second BaaS restart. Degrading to
+        # "no poll for this restart" (its completion falls back to
+        # /restart_status, the behaviour before this task existed) is strictly
+        # better than re-deploying the bot.
+        try:
+            enqueue_restart_poll(self._task_queue_service, publish_id=publish_id)
+        except Exception as exc:
+            logger.warning(
```
[the remark this answers is not in the document]
Comment (totalfrank) [reply src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py]: You're right, and the rule you cited settles it — swallowing the write to dodge the redelivery was the wrong trade. Fixed properly in `98601c6` by taking your first option: make redelivery recognise the already-submitted workflow, then let the enqueue propagate.

`has_unreconciled_restart(publish_id, stage)` gates the handler. It requires all three of:

1. `ext.restart.restarting` set,
2. `ext.restart.<stage>` holding a workflow handle,
3. **no deploy operation still open for the stage** — neither RESTART nor FIRST_RELEASE in a non-terminal state.

The third is the load-bearing one. A crash mid-`execute_restart` leaves its RESTART op open, and a crash between a recreate's ext write and its `complete_operation` leaves a FIRST_RELEASE op open — both must fall *through* the guard so `acquire_deploy_workflow` resumes by adopt-by-query and `_finalize_dangling_recreate_op` finalises the dangling leg, exactly as they do today. Only a stage with nothing left open is genuinely "submitted, nothing to run but the observation".

With that in place the `try/except` is gone and `enqueue_restart_poll` raises freely. A failed write now retries the task, which hits the guard and re-enqueues only the poll — which is what you asked for, and what `AGENTS.md:264-265` requires.

Consequence worth naming: a second `/restart` arriving while one is genuinely in flight is now absorbed as a redelivery rather than issuing a second restart. The endpoint's response is unchanged, and I think that is the better behaviour — but it is a real semantic change, so it is called out in the PR body rather than left implicit.

Three tests: the enqueue failure propagates, a redelivery enqueues the poll without calling `execute_restart`, and a first delivery still issues the restart.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by totalfrank** [comment]
## Correction: CI does run on this PR

My earlier comment on this thread said no GitHub Actions run had been produced for this PR and speculated about the cause. **That was wrong, and I've removed the claim from the PR description.** The run simply took longer to appear than I allowed for; `Backend unit tests` executed against `717b4e2` and reported in ~10 minutes.

Apologies to anyone who read that as "CI is broken here" — it isn't.

## What CI caught

```
FAILED tests/community/architecture/test_no_oversized_modules.py::test_no_new_oversized_files
  core/service_bot/services/publish_flow_service.py: 1006 lines (cap 1000)
```

`10361 passed, 1 failed`. A genuine gate, and one every local run I did had missed — my selections were scoped to `core/service_bot`, `core/task_queue` and two endpoint files, so `tests/community/architecture/` was never included. Corrected: I now run the full suite before pushing.

Fixed in `98601c6` by moving both task-handler accessors to `restart_mixin.py`, where the restart concern already lives and which the service composes. `publish_flow_service.py` returns to **988 lines** — exactly its length on `REL20260804`.

## Round 4 findings

| # | Finding | Outcome |
|---|---------|---------|
| 6 | P1 — the round-3 fix swallowed a failed persistence write, against `AGENTS.md:264-265` | **Fixed** `98601c6` — redelivery guard added, `try/except` removed, write propagates |
| 7 | P2 — a record superseded to `UPGRADED` mid-restart stops being polled | **Deferred → #784**, with a correction to the diagnosis |

Finding 6 was a fair hit on my round-3 patch: I had traded a rule violation for a bug fix. The proper fix is `has_unreconciled_restart`, which lets the handler recognise a restart it already submitted and re-run only the poll handoff — so the write can propagate as the rule requires, and redelivery still cannot issue a second BaaS restart.

**One behaviour change from that, flagged for review:** a second `/restart` arriving while one is genuinely in flight is now absorbed as a redelivery rather than issuing a second restart. The endpoint's response is unchanged. I think it's the better behaviour, but it's a real change and shouldn't pass unnoticed.

On finding 7, the guard isn't the cause — `sync_restart_progress` derives its stage from the publish status, so an `UPGRADED` record is unreconcilable by it either way; removing the guard would trade an immediate exit for a 10,800-iteration spin to the 24h deadline. It's the most reachable of the three deferrals, since it needs only one restart racing the next release rather than overlapping restarts. Filed as its own item on #784.

## State

Full backend suite locally: **10359 passed, 3 skipped** — the same 10,362 total CI ran, with its one failure fixed. Architecture gates pass (112). All seven review threads answered and resolved.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by xianmuyq — APPROVED** [review]

File: src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py [hunk 7]
```diff
@@ -370,12 +390,44 @@ def handle(self, payload: Optional[dict]) -> TaskOutcome:
...
 
     async def _run(self, publish_id: int, stage: str, operator: str) -> TaskOutcome:
+        # Redelivery guard. The queue is at-least-once and ``execute_restart``
+        # COMPLETEs its ledger op before returning, so past that point a second
+        # delivery would open the next attempt and issue a second BaaS restart.
+        # When the record already carries a submitted-but-unreconciled restart
+        # for this stage, re-run only the handoff. That is what lets the enqueue
+        # below propagate its failures instead of swallowing them: the task
+        # retries, lands here, and re-enqueues the poll without re-deploying.
+        if self._flow.has_unreconciled_restart(publish_id, stage):
```
[the remark this answers is not in the document]
Comment (totalfrank) [reply src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/tasks.py:400]: Correct, and it's a sharper statement of the guard's limit than the one I wrote — my PR body claims the guard covers the crash window "whenever the marker and handle are already written", and your sequence is precisely a case where they *were* written and the guard still misses. I've corrected that wording.

Walking it through to confirm: delivery 1 submits, completes the op, writes marker + handle, enqueues the poll; the pod dies before the queue marks the row SUCCEEDED; the poll (already a durable row of its own) runs, sees a fast terminal workflow, and `_handle_sync_success` pops `restarting`; the lease expires and delivery 2 finds no marker, so `has_unreconciled_restart` is false and `execute_restart` opens attempt N+1 against a terminal RESTART op. Duplicate BaaS restart. The diagnosis is right: I used a transient reconciliation marker as a delivery identity, and its lifetime is shorter than the delivery's.

Deferring to **#784**, where this is hazard 2 — and it sharpens that issue rather than adding to it, because it rules out the cheap patch. No refinement of the marker heuristic can work: the only signal that distinguishes "redelivery of task X" from "a genuine later restart request" is an identity carried by the delivery itself. That is the `restart_bot`-opens-the-operation / op-id-in-the-payload fix already sketched there, and it is a change to the restart entry point.

What the guard does buy, stated honestly: redelivery re-issues only when the marker has already been cleared, instead of always. Strictly better than before it existed, and strictly short of correct.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/retry_ops_mixin.py [hunk 8]
```diff
@@ -208,9 +209,15 @@ async def _retry_via_restart(
...
-            # the restart is settled by the user-called /restart_status endpoint.
             enqueue_progress_poll(self._task_queue_service, publish_id=publish_id)
+            # SUCCESS is not a poll wait state, so the poll above completes
+            # immediately and would leave this restart depending on a client
+            # calling /restart_status — the very gap the restart poll closes. It
+            # keys on ext.restart.restarting (set by the execute_restart above)
+            # rather than the status, so it covers both rollback targets. For the
+            # *_PUB ones that means two pollers on one sync; they are idempotent
+            # and whichever observes the terminal workflow first ends both.
+            enqueue_restart_poll(self._task_queue_service, publish_id=publish_id)
```
[the remark this answers is not in the document]
Comment (totalfrank) [reply src/backend/src/agentclaw/community/core/service_bot/services/publish_flow/retry_ops_mixin.py:220]: Real, and the "no restart task exists to retry this handoff" observation is the part that makes it distinct — the retry path runs `execute_restart` inline precisely so the poll can't be claimed in the same worker batch, which also means there is no durable row to redeliver. Combined with the record having left `FAILED`, a failed enqueue there is genuinely unrecoverable without `/restart_status`.

Two things frame what I'm doing about it.

**It is not a regression.** Before this branch, the `SUCCESS`-target retry path depended on `/restart_status` unconditionally — that was your round-2 finding. This change makes it durable on the success path and leaves it exactly at the old behaviour when the enqueue write fails. So the worst case here equals the previous every case.

**It is systemic, not local to this line.** Every handoff in this flow is a one-shot enqueue with the same property: `_retry_via_online_release` advances the record to `ONLINE_PUB` and then enqueues `online_release`, `_retry_via_verify_flow` does the same for `verify_flow`, and `process` does it for the initial dispatch. A failed write strands the record identically in each. Making this one atomic without the others would be a local patch on a general property, and doing it properly means a transactional outbox — the enqueue committing in the same transaction as the state change it hands off.

So: recorded on **#784** rather than patched here. One cheap improvement I considered and also left for that work: for a `SUCCESS` rollback target the `progress_poll` enqueue is provably a no-op (the handler completes immediately on a non-`*_PUB` status), so that path currently spends its *first* write on a task that does nothing and its second on the one that matters. Enqueueing only the restart poll for that target would both remove a dead row and put the useful write first. Small, and it belongs with the durability work rather than as a standalone change to a path already under review.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by totalfrank** [comment]
## Review round 5 — two P1s landed, both deferred

@xianmuyq — flagging this because your approval and these findings arrived within a minute of each other, so you may not have seen them. **No code changed in response**, so the approval still describes the diff you reviewed. But two P1s were raised and I deferred both to #784 rather than fixing them, and that's a judgement you should get to overrule.

| # | Finding | Outcome |
|---|---------|---------|
| 8 | P1 — redelivery can still duplicate a restart once the poll has cleared the marker | **Deferred → #784** |
| 9 | P1 — the `SUCCESS`-retry poll handoff is a one-shot write with no durable row to retry | **Deferred → #784** |

### Why deferred

**Finding 8** is the sharpest result of the whole review, and it invalidates the cheap fix rather than asking for one. `has_unreconciled_restart` keys on `ext.restart.restarting` — a *reconciliation* marker whose lifetime is shorter than a *delivery's*. If the pod dies after the poll is enqueued but before the task row is marked SUCCEEDED, the poll can observe a fast terminal workflow and clear the marker first; the redelivered task then finds no marker and issues a second BaaS restart. No refinement of that heuristic can work, because once the first restart settles a redelivery and a genuine new request look identical. The discriminator has to be a delivery-carried operation identity — `restart_bot` opening the ledger op and putting its id in the payload — which changes the restart entry point.

I have corrected the PR body, which previously claimed the guard covers the crash window "whenever the marker and handle are already written". It does not, and finding 8 is exactly the counterexample.

**Finding 9** is real but systemic rather than local: every handoff in this flow is a one-shot enqueue that can strand a record if it fails — `_retry_via_online_release`, `_retry_via_verify_flow` and `process` all share the shape. The general fix is a transactional outbox. Patching this one call site would be a local fix to a general property.

### What this branch still buys

Neither finding is a regression, and both failure modes are strictly narrower than before:

- redelivery re-issues only once the marker has been cleared, instead of always;
- the `SUCCESS`-retry path is now durable on the success path, where before it depended on `/restart_status` unconditionally.

The single-restart correctness this PR is about is unaffected. If you'd rather the entry-point fix land here instead of in #784 — which would resolve findings 8, 5 and 2's residual at once, at the cost of `/restart` refusing a request while one is in flight — say so and I'll do it.

### CI

5 of 7 checks green on `98601c6`; `Backend unit tests` and `Singlebox coverage` still running. I'll report if either goes red.

---
_Generated by [Claude Code](https://claude.ai/code)_
